### PR TITLE
CRM-20220 fix signature fields to be longtext

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
* [CRM-20220: Use text box for signature field on payment processor page](https://issues.civicrm.org/jira/browse/CRM-20220)

### CODE REVIEW

File: CRM/Admin/Form/PaymentProcessor.php [hunk 1]
```diff
@@ -44,15 +44,8 @@ class CRM_Admin_Form_PaymentProcessor extends CRM_Admin_Form {
   protected $_ppDAO;
 
   public function preProcess() {
-    if (!CRM_Core_Permission::check('administer payment processors')) {
```
Comment (eileenmcnaughton) [inline CRM/Admin/Form/PaymentProcessor.php]: moved this to the xml definition of the form

File: CRM/Admin/Form/PaymentProcessor.php [hunk 2]
```diff
@@ -44,15 +44,8 @@ class CRM_Admin_Form_PaymentProcessor extends CRM_Admin_Form {
...
-      CRM_Core_Error::statusBounce('The \'administer payment processors\' permission is required to add or edit a payment processor.');
-    }
     parent::preProcess();
 
-    CRM_Utils_System::setTitle(ts('Settings - Payment Processor'));
```
Comment (eileenmcnaughton) [inline CRM/Admin/Form/PaymentProcessor.php]: ditto

File: CRM/Admin/Form/PaymentProcessor.php [hunk 3]
```diff
@@ -77,10 +70,6 @@ public function preProcess() {
     $this->_ppDAO = new CRM_Financial_DAO_PaymentProcessorType();
     $this->_ppDAO->id = $this->_ppType;
 
-    if (!$this->_ppDAO->find(TRUE)) {
```
Comment (eileenmcnaughton) [inline CRM/Admin/Form/PaymentProcessor.php]: this is checking for an impossible condition

File: CRM/Core/Form.php [hunk 4]
```diff
@@ -1375,7 +1379,7 @@ public function addSelect($name, $props = array(), $required = FALSE) {
    * @throws \Exception
    * @return HTML_QuickForm_Element
    */
-  public function addField($name, $props = array(), $required = FALSE, $legacyDate = TRUE) {
+  public function addField($name, $props = array(), $required = FALSE, $fieldSpec = array(), $legacyDate = TRUE) {
```
Comment (eileenmcnaughton) [inline CRM/Core/Form.php]: @colemanw what is your opinion on this. It feels like the easiest way to make the fields on the payment processor form metadata driven. However, some of the fields are pseudofields so I figured maybe we could change the scope here to allow them to be passed in

File: CRM/Contact/Form/Edit/Demographics.php [hunk 5]
```diff
@@ -45,10 +45,10 @@ class CRM_Contact_Form_Edit_Demographics {
   public static function buildQuickForm(&$form) {
     $form->addField('gender_id', array('entity' => 'contact', 'type' => 'Radio', 'allowClear' => TRUE));
 
-    $form->addField('birth_date', array('entity' => 'contact'), FALSE, FALSE);
+    $form->addField('birth_date', array('entity' => 'contact'), FALSE, array(), FALSE);
```
Comment (eileenmcnaughton) [inline CRM/Contact/Form/Edit/Demographics.php]: The legacyDateParam is v recent & transitional so I know these are the only fields using it

**Comment by colemanw** [comment]
@eileenmcnaughton unfortunately this is stale

File: CRM/Core/Form.php (hunk 4 above)
Comment (colemanw) [inline CRM/Core/Form.php]: I don't think we ought to use this function with pseudofields.

**Comment by eileenmcnaughton** [comment]
@colemanw I fixed the staleness. Re adding the test fields: here's the thing, the fields create a second row in the payment processor table, but they arguably are 'real fields' because the way it works is you update them in practice as if they were one thing (I might almost wish they were one row in the table!)

I did think about adding them to getfields so the api understood them as part of the set of fields to be added, if we were thinking about putting them in one row, one day that would be a good step towards that.

The difficulty would be on the off chance someone really did want to use the api to alter just the test fields, hence I thought maybe I should do trickery at the form layer.

I guess I can go back to 'add' if there is no metadata way to do this

**Comment by colemanw** [comment]
I agree with `add`. The `addField` function has its limitations and was never intended to work with fake fields.

**Comment by eileenmcnaughton** [comment]
@colemanw ok, I've switched the test fields back...

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Admin/Form/PaymentProcessor.php [hunk 6]
```diff
@@ -43,16 +43,21 @@ class CRM_Admin_Form_PaymentProcessor extends CRM_Admin_Form {
 
   protected $_ppDAO;
 
+  /**
+   * Get the base entity being edited.
+   *
+   * @return string
+   *   The entity being edited.
+   *
+   * @throws Exception
```
Comment (colemanw) [inline CRM/Admin/Form/PaymentProcessor.php]: no it doesn't

**Comment by eileenmcnaughton** [comment]
@colemanw this has passed tests with your update, do you think it's good to merge now?

**Pull request merged**